src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs

```diff
@@ -3241,6 +3241,369 @@ class Q: System.Runtime.CompilerServices.ITuple
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+    }
+}
+struct S
+{
+    public int N;
+    public int Q => N;
+
+    public S(int n) => N = n;
+
+    public bool No() { N++; return false; }
+    public bool Nope { get { N++; return false; } }
+    public bool this[int t] { get { N++; return false; } }
+}
+";
+            var expectedOutput = "222222";
+            var compilation = CreateCompilation(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics(
+                );
+            CompileAndVerify(compilation, expectedOutput: expectedOutput);
+        }
+
+        [Fact]
+        public void DoNotShareTempMutatedThroughReceiver_02()
+        {
+            var source = @"
+using System;
+class Program
+{
+    static void Main()
+    {
+        M<S>(new S(1));
+    }
+
+    static void M<S>(S news) where S : I
+    {
+        S s;
+        s = copy();
+        Console.Write(s switch
+        {
+            { N: 1 } when s.No() => 1,
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+
+        Console.Write(copy() switch
+        {
+            { N: 1 } s0 when s0.No() => 1,
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+
+        s = copy();
+        Console.Write(s switch
+        {
+            { N: 1 } when s.Nope => 1,
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+
+        Console.Write(copy() switch
+        {
+            { N: 1 } s0 when s0.Nope => 1,
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+
+        s = copy();
+        Console.Write(s switch
+        {
+            { N: 1 } when s[0] => 1,
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+
+        Console.Write(copy() switch
+        {
+            { N: 1 } s0 when s0[0] => 1,
+            { Q: 1 } => 2,
+            _ => 3,
+        });
+
+        S copy() => news;
+    }
+}
+interface I
+{
+    int N { get; }
+    int Q { get; }
+
+    bool No();
+    bool Nope { get; }
+    bool this[int t] { get; }
+}
+struct S : I
+{
+    public int N { get; private set; }
+    public int Q => N;
+
+    public S(int n) => N = n;
+
+    public bool No() { N++; return false; }
+    public bool Nope { get { N++; return false; } }
+    public bool this[int t] { get { N++; return false; } }
+}
+";
+            var expectedOutput = "222222";
+            var compilation = CreateCompilation(source, options: TestOptions.DebugExe);
+            compilation.VerifyDiagnostics(
+                );
+            CompileAndVerify(compilation, expectedOutput: expectedOutput);
+        }
+
+        [Fact]
+        public void DoNotShareTempMutatedThroughLambda_01()
+        {
+            var source = @"
+using System;
+class Program
+{
+    static void Main()
+    {
+        S s = new S(1);
+        Func<bool> condition = () => { s = new S(10); return false; };
+        Console.Write(s switch
+        {
+            { N: 1 } when condition() => 1,
```

Comment:
>condition() [](start = 26, length = 11)

It doesn't look like this delegate modifies a pattern variable. Isn't that what WhenClauseMightAssignPatternVariableWalker is supposed to prevent?

Reply:
Such a lambda would have to appear in a when clause and contain an assignment. That is blocked by `VisitAssignment`. And we do have tests for assignments.

---
In reply to: [406986224](https://github.com/dotnet/roslyn/pull/42972#discussion_r406986224) [](ancestors = 406986224)

Reply:
> Such a lambda would have to appear in a when clause and contain an assignment. That is blocked by VisitAssignment.

Let's have a test like that and observe the fact.
Let's also add a test with a delegate backed by a local function that mutates a pattern local (can be done in a switch statement).


---
In reply to: [407003266](https://github.com/dotnet/roslyn/pull/42972#discussion_r407003266) [](ancestors = 407003266,406986224)